### DISCUSSION TITLE
Rules being checked after BO Save and triggering PropertyChanged

### DISCUSSION THREAD

**Discussion by mtavares628** [MarimerLLC/csla#2827] [Questions] [answered]
It looks like my validation rules are running again after the save of my business object is complete.  I'm trying to figure out if this is something I'm inadvertently doing somewhere, or if this is by design.  So, does BusinessBase run CheckRules again after an update?  Further, it looks like some of my business rules are triggering OnPropertyChanged for the PrimaryProperty of the rule. Is this normal?  I don't appear to be actually changing the property. Here's the rule:

```c#
private class LivingArrangementRequiresExplanation : CommonRules.NetsoftCommonBusinessRuleAsync
        {
            private IPropertyInfo AdvLivingArrangementIDProperty;

            public LivingArrangementRequiresExplanation(IPropertyInfo primaryProp, IPropertyInfo AdvLivingArrangementIDProp) : base(primaryProp)
            {
                AdvLivingArrangementIDProperty = AdvLivingArrangementIDProp;
                InputProperties = new List<IPropertyInfo>() { PrimaryProperty, AdvLivingArrangementIDProperty };
            }

            [System.Diagnostics.CodeAnalysis.SuppressMessage("Usage", "CSLA0017:Find Business Rules That Do Not Use Add() Methods on the Context", Justification = "<Pending>")]
            protected override async Task ExecuteAsync(IRuleContext context)
            {
                int advLivingArrangmentID = context.GetInputValue<int>(AdvLivingArrangementIDProperty);
                string LivingArrangementExplanation = context.GetInputValue<string>(PrimaryProperty);
                if (string.IsNullOrEmpty(LivingArrangementExplanation))
                {
                    var cmd = await DataPortal.CreateAsync<LivingArrangementRequiresExplanationCommand>(advLivingArrangmentID, NdrnCommon.GetUserAdvocacyGroup());
                    cmd = await DataPortal.ExecuteAsync<LivingArrangementRequiresExplanationCommand>(cmd);
                    if (cmd is object && cmd.RequiresExplanation)
                    {
                        string desc;
                        if (string.IsNullOrEmpty(GetMessage()))
                        {
                            desc = "The Living Arrangement selected requires an explanation";
                        }
                        else
                        {
                            desc = MessageText;
                        }
                        SetContext(context, desc);
                    }
                }
            }

            protected void SetContext(Csla.Rules.IRuleContext context, string desc)
        {
            SetContext(context, desc, Severity);
        }

        protected void SetContext(Csla.Rules.IRuleContext context, string desc, RuleSeverity sev)
        {
            switch (sev)
            {
                case RuleSeverity.Error:
                    {
                        context.AddErrorResult(desc);
                        break;
                    }

                case RuleSeverity.Information:
                    {
                        context.AddInformationResult(desc);
                        break;
                    }

                case RuleSeverity.Warning:
                    {
                        context.AddWarningResult(desc);
                        break;
                    }

                case RuleSeverity.Success:
                    {
                        context.AddSuccessResult(false);
                        break;
                    }
            }
        }
    }
        }
```

I'm not calling AddOutValue on the context, so I don't see why it would end up calling OnPropertyChanged for my ExplanationProperty.  But if were to comment out this rule from running, then the PropertyChanged event isn't raised for my ExplanationProperty.    This is happening with several rules that are hitting my database via a Command Object, but I don't see how that would generate a PropertyChanged even either. Any ideas?

**Comment by rockfordlhotka** [comment]
> it looks like some of my business rules are triggering OnPropertyChanged for the PrimaryProperty of the rule

This is expected, because we don't have a way of knowing whether the UI needs refreshing based on the result of the rule (not the property value itself, but whether the UI is showing an info/warn/error message that needs to be updated).

**Comment by rockfordlhotka** [comment, marked as answer]
> does BusinessBase run CheckRules again after an update?

Interesting. I created a simple test, and `SaveAsync` doesn't, but `SaveAndMergeAsync` does.

If I remember correctly, this is intended behavior, because we're merging the result of the update back into the original object graph, and the save operation may have changed property values on the logical server. As we merge those (potentially changed) property values back into the original object graph, it is necessary to rerun rules because the (potentially) new values might be different.

**Reply by mtavares628** [reply]
It looks like the SaveAsync on the ViewModel (for blazor) is calling some kind of MergeGraph after the SaveAsync. Is that the same functionality as what the SaveAndMergeAsync does?  If so that answers my question.

Here is the code from ViewModel:

```c#
/// <summary>
    /// Override to provide custom Model save behavior
    /// </summary>
    /// <returns></returns>
    protected virtual async Task<T> DoSaveAsync()
    {
      if (Model is Core.ISavable savable)
      {
        var result = (T)await savable.SaveAsync();
        if (Model is Core.IEditableBusinessObject editable)
          new Core.GraphMerger().MergeGraph(editable, (Core.IEditableBusinessObject)result);
        else
          Model = result;
      }
      return Model;
    }
```

**Reply by rockfordlhotka** [reply]
Yes, exactly, `SaveAndMergeAsync` calls the `GraphMerger` to merge the result from the server back into the existing client-side object graph.

**Reply by rockfordlhotka** [reply]
It is really a multi-part choice. Not only do we (probably) want the rules to run. But also any property values that actually changed need to raise `PropertyChanged` events so the UI updates accordingly.

**Reply by mtavares628** [reply]
Got it, thanks.  It makes sense, although there are definitely cases where I know that the rules don't need to run again coming back from the save.  I wonder if an overload of the SaveAsync on the ViewModel that would bypass the graph merge might make sense for those times where performance matters and you know that you don't need to run the rules again until the user changes the properties manually. Maybe not, I don't know. I'm just worried that if some of my other BOs that have a lot of validation rules that hit the database end up re-running all of the rules again once they come back from the save, it's going to bog down the UI a bit. Especially, since they will potentially run the rule an extra time when the PropertyChanged event is raised for a dependent property.

**Reply by rockfordlhotka** [reply]
You can certainly override the virtual method - copy-paste the existing code and wrap the `GraphMerger` call in code that uses the `ICheckRules` interface, which allows you to suppress and resume rule checking.

**Reply by mtavares628** [reply]
So suppressing the rule checking on the parent object only partially worked for me.  All of the child objects still ran their business rules when the GraphMerger was called.  I'm guessing that because of the recursive nature of that MergeGraph method, the child object check rules are being called, and I haven't set SuppressRuleChecking on the child objects so they are still being run.  Here's my overridden method. Let me know if I'm not implementing your suggestion correctly:

```c#
protected override async Task<T> DoSaveAsync()
    {
        if (Model is Csla.Core.ISavable savable)
        {
            var result = (T)await savable.SaveAsync();
            if (Model is Csla.Core.IEditableBusinessObject editable)
            {
                ((Csla.Core.ICheckRules)editable).SuppressRuleChecking();
                new Csla.Core.GraphMerger().MergeGraph(editable, (Csla.Core.IEditableBusinessObject)result);
                ((Csla.Core.ICheckRules)editable).ResumeRuleChecking();
            }
            else
                Model = result;
        }
        return Model;
    }
```
Is there a way to cascade the rule suppression to the child objects?

**Reply by rockfordlhotka** [reply]
I think you'll need something like this:

```c#
        if (obj is IManageProperties target)
          foreach (var item in target.GetChildren())
            // suppress for item
```

**Reply by mtavares628** [reply]
@rockfordlhotka  it looks like IManageProperties is internal to the assembly, so I can't use it in my override.

**Reply by rockfordlhotka** [reply]
Oh, darn! You might want to create a type to do this work, so you can subclass `ObjectFactory` and make your life a bit easier. Sort of like:

```c#
using Csla.Core;

namespace Test
{
  public class GraphSuppressor : Csla.Server.ObjectFactory
  {
    public void SuppressRules(object obj)
    {
      if (obj is ICheckRules host)
        host.SuppressRuleChecking();
      var properties = Csla.Core.FieldManager.PropertyInfoManager.GetRegisteredProperties(obj.GetType());
      foreach (var property in properties)
        if (property.IsChild)
          SuppressRules(ReadProperty(obj, property));
    }
  }
}
```

**Reply by mtavares628** [reply]
Thanks Rocky. Unfortunately, there is no IsChild property on IPropertyInfo, I tried checking whether the property type is IEditableBusinessObject or IEditableCollection.  However, it still wasn't suppressing for the child collections, so I think I'm missing that extra step to loop through the collection and suppress the rules for each child object. My issue now is that I'm not sure how to do that properly in a generic way. Here's what I currently have:

```c#
public class GraphSuppressor : Csla.Server.ObjectFactory
    {
        public void SuppressRules(object obj)
        {
            if (obj is ICheckRules host)
                host.SuppressRuleChecking();
            else if (obj is IEditableCollection childList)
            {
                // not sure what type to cast this to here to loop through on each object
                // since IEditableCollection doesn't implement IEnumerable
            }
            var properties = Csla.Core.FieldManager.PropertyInfoManager.GetRegisteredProperties(obj.GetType());
            foreach (var property in properties)
                if (property.Type is IEditableBusinessObject || property.Type is IEditableCollection)
                    SuppressRules(ReadProperty(obj, property));
        }
    }
```
What should I cast the collection to so that I can loop through the child objects?

**Reply by mtavares628** [reply]
Nevermind, I figured it out.  I don't know why it didn't occur to me to just cast as IEnumerable itself to perform the loop. Here is what I ended up with:

```c#
public class GraphSuppressor : Csla.Server.ObjectFactory
    {
        public void SuppressRules(object obj)
        {
            if (obj is ICheckRules host)
                host.SuppressRuleChecking();
            else if (obj is IEditableCollection childList)
            {
                foreach (var item in (IEnumerable)childList)
                {
                    SuppressRules(item);
                }
            }
            var properties = Csla.Core.FieldManager.PropertyInfoManager.GetRegisteredProperties(obj.GetType());
            foreach (var property in properties)
            {
                var prop = ReadProperty(obj, property);
                if (prop is IEditableBusinessObject || prop is IEditableCollection)
                    SuppressRules(prop);
            }
                
        }

        public void ResumeRules(object obj)
        {
            if (obj is ICheckRules host)
                host.ResumeRuleChecking();
            else if (obj is IEditableCollection childList)
            {
                foreach (var item in (IEnumerable)childList)
                {
                    ResumeRules(item);
                }
            }
            var properties = Csla.Core.FieldManager.PropertyInfoManager.GetRegisteredProperties(obj.GetType());
            foreach (var property in properties)
            {
                var prop = ReadProperty(obj, property);
                if (prop is IEditableBusinessObject || prop is IEditableCollection)
                    ResumeRules(prop);
            }
        }
    }
```

**Reply by michaelcsikos** [reply]
Hi @rockfordlhotka, we have a situation where two child objects are validated by the parent. When using `SaveAndMergeAsync` this can fail as the second child object hasn't had its values merged in yet. Apart from avoiding `SaveAndMergeAsync`, is there something we can do to solve this?
